### PR TITLE
Created models and views to support adding new contest

### PR DESCRIPTION
**Pull request by abassawo**
WIP - But I'd like a review in any event, because a lot of the contest model is defined here. Most of the view stuff is subject to change. In particular, as discussed with @jmccormack200 I'm planning on breaking down the NewContestPresenter's responsibility and pairing each fragment in the contestcreation package with a presenter of its own.

### CODE REVIEW

File: app/src/main/java/io/intrepid/contest/BaseSlideFragment.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import butterknife.ButterKnife;
+import butterknife.Unbinder;
+import io.intrepid.contest.base.BaseFragment;
+import timber.log.Timber;
+
+public abstract class BaseSlideFragment extends Fragment {
+
+    public abstract boolean canMoveFurther();
+    public abstract int cantMoveFurtherErrorMessage();
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/BaseSlideFragment.java]: It's not very clear that this is a resource id, maybe you can return a String and inside the method get the actual value with getResources.getString(R.id._____)

File: app/src/main/java/io/intrepid/contest/base/BaseFragmentActivity.java [hunk 2]
```diff
@@ -24,12 +24,16 @@ protected void onCreate(Bundle savedInstanceState) {
         if (savedInstanceState == null) {
             Intent intent = getIntent();
             Fragment fragment = createFragment(intent);
-            getSupportFragmentManager()
-                    .beginTransaction()
-                    .add(R.id.fragment_container, fragment)
-                    .commit();
+            hostFragment(fragment);
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/base/BaseFragmentActivity.java]: Why did you move it if it's not used anywhere else?

File: app/src/main/java/io/intrepid/contest/base/BaseMvpActivity.java [hunk 3]
```diff
@@ -25,7 +25,7 @@
      * Override {@link #onViewCreated(Bundle)} to handle any logic that needs to occur right after inflating the view.
      * onViewCreated is called immediately after onCreateView
      */
-    protected final void onCreate(@Nullable Bundle savedInstanceState) {
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/base/BaseMvpActivity.java]: Why did you make it not final?

File: app/src/main/java/io/intrepid/contest/base/BasePresenter.java [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 import android.support.annotation.NonNull;
 
+import io.intrepid.contest.models.Category;
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/base/BasePresenter.java]: You probably pulled methods to BasePresenter by mistake and this is leftover code.

File: app/src/main/java/io/intrepid/contest/models/User.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
+package io.intrepid.contest.models;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+/*{
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/models/User.java]: We could create a standard JavaDoc "template" to use in our API models. Not sure if Retrofit has any standards.

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        return super.onOptionsItemSelected(item);
+    }
+
+    @Override
+    public int getNumberOfScreens() {
+        return NUMBER_OF_SCREENS;
+    }
+
+    public void acceptContestName(String contestName) {
+        presenter.setContestName(contestName);
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java]: Other presenters' methods are in standard "onContestNameAccepted"

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Override
+    public void acceptContestDescription(String contestDescription) {
+        presenter.setContestDescription(contestDescription);
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java]: [Same] Other presenters' methods are in standard "onContestDescriptionAccepted"

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Override
+    public void acceptCategory(Category category) {
+        presenter.addCategory(category);
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java]: [Same] Other presenters' methods are in standard "onContestCategoryAccepted"

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Override
+    public void onCancelClicked() {
+        startActivity(SplashActivity.getIntent(this)); //fixme - NavUtil or onBackPressed - no ned for new intent.
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java]: I used makeIntent() instead of getIntent, we can decide in one of them and standardize our methods afterwards.

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/describecontest/DescribeContestFragment.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+import butterknife.BindView;
+import butterknife.OnTextChanged;
+import io.intrepid.contest.R;
+import io.intrepid.contest.screens.contestcreation.BaseNewContestFragment;
+import io.intrepid.contest.screens.contestcreation.NewContestActivity;
+
+public class DescribeContestFragment extends BaseNewContestFragment {
+    private static final String CONTEST_NAME = "ContestName";
+    @BindView(R.id.contest_description_edittext)
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/describecontest/DescribeContestFragment.java]: I like to add a blank line between Butterknife variables and the others, not sure if you meant to do the same or if you don't care for it (which is totally fine)

Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/describecontest/DescribeContestFragment.java]: Also I'm doing edit_text, text_view etc.

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/namecontest/NameContestContract.java [hunk 11]
```diff
@@ -0,0 +1,4 @@
+package io.intrepid.contest.screens.contestcreation.namecontest;
+
+public class NameContestContract {
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/namecontest/NameContestContract.java]: I think this will change when you create presenters for your fragments, but did you mean not to use NameContestContract to abstract your Presenter/Activity?

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/namecontest/NameContestFragment.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+import butterknife.OnTextChanged;
+import io.intrepid.contest.R;
+import io.intrepid.contest.screens.contestcreation.BaseNewContestFragment;
+import io.intrepid.contest.screens.contestcreation.NewContestActivity;
+
+public class NameContestFragment extends BaseNewContestFragment {
+    @BindView(R.id.contest_name_edittext)
+    EditText contestNameField;
+
+
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/namecontest/NameContestFragment.java]: Extra blank line

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/namecontest/NameContestFragment.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+import android.widget.EditText;
+
+import butterknife.BindView;
+import butterknife.OnTextChanged;
+import io.intrepid.contest.R;
+import io.intrepid.contest.screens.contestcreation.BaseNewContestFragment;
+import io.intrepid.contest.screens.contestcreation.NewContestActivity;
+
+public class NameContestFragment extends BaseNewContestFragment {
+    @BindView(R.id.contest_name_edittext)
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/namecontest/NameContestFragment.java]: [Same] I'm using edit_text, text_view etc.

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/namecontest/NameContestPresenter.java [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+import io.intrepid.contest.base.BaseContract;
+
+public class NameContestPresenter {
+
+    public interface View extends BaseContract.View{
+        void acceptContestName(String contestName);
+    }
+
+    public interface Presenter {
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/namecontest/NameContestPresenter.java]: The presenter has a Presenter interface? Maybe you could rename this to NameContestContract, which is empty?
(But again, you said this stuff will probably change)

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/reviewcontest/CategoriesListFragment.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+import io.intrepid.contest.R;
+import io.intrepid.contest.screens.contestcreation.BaseNewContestFragment;
+
+public class CategoriesListFragment extends BaseNewContestFragment {
+
+    @Override
+    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
+        super.onCreateOptionsMenu(menu, inflater);
+        MenuItem nextButton = menu.findItem(R.id.action_next);
+        nextButton.setIcon( android.R.drawable.ic_media_next);
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/reviewcontest/CategoriesListFragment.java]: Extra space only in one side inside parentheses

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/reviewcontest/CategoriesListFragment.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    @Override
+    protected int getLayoutResourceId() {
+        return R.layout.fragment_contest_categories_rv;
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/reviewcontest/CategoriesListFragment.java]: Why does the name end in "rv"?

File: app/src/main/java/io/intrepid/contest/utils/SlidingTabAdapter.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+import java.util.ArrayList;
+
+import io.intrepid.contest.BaseSlideFragment;
+import timber.log.Timber;
+
+public class SlidingTabAdapter extends FragmentPagerAdapter implements ViewPager.OnPageChangeListener{
+    private final AppCompatActivity activity;
+    private int currentIndex = 0;
+    private ArrayList<BaseSlideFragment> fragments = new ArrayList<>();
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/utils/SlidingTabAdapter.java]: Probably best to be instantiated somewhere else

File: app/src/main/java/io/intrepid/contest/utils/SlidingTabAdapter.java [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+        currentIndex = position;
+        Timber.d("Current index " + currentIndex);
+    }
+
+    @Override
+    public void onPageSelected(int position) {
+        currentIndex = position;
+        Timber.d("Current index " + currentIndex);
+        switch (currentIndex){
+            case 0: activity.getSupportActionBar().setTitle("New Title");
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/utils/SlidingTabAdapter.java]: Strings can be extracted to resources

File: app/src/main/java/io/intrepid/contest/utils/SlidingTabAdapter.java [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+                break;
+            case 3: activity.getSupportActionBar().setTitle("Scoring Categories");
+                break;
+        }
+
+    }
+
+    @Override
+    public void onPageScrollStateChanged(int state) {
+
```
Comment (gsintrepid) [inline app/src/main/java/io/intrepid/contest/utils/SlidingTabAdapter.java]: Is it mandatory to implement it?

File: app/src/main/res/layout/activity_fragment_container.xml [hunk 20]
```diff
@@ -1,7 +1,31 @@
 <?xml version="1.0" encoding="utf-8"?>
-<FrameLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
-    android:id="@+id/fragment_container"
-    android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    />
+<android.support.design.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (gsintrepid) [inline app/src/main/res/layout/activity_fragment_container.xml]: Why did you change back to CoordinatorLayout, for the appbar?

File: app/src/main/res/layout/activity_viewpager_container.xml [hunk 21]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.design.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (gsintrepid) [inline app/src/main/res/layout/activity_viewpager_container.xml]: [Same] Why did you change back to CoordinatorLayout, for the appbar?

File: app/src/main/res/layout/fragment_edit_contest_category.xml [hunk 22]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+              android:layout_width="match_parent"
+              android:layout_height="match_parent"
+              android:layout_margin="@dimen/activity_vertical_margin"
+              android:orientation="vertical"
+    >
+
+    <android.support.design.widget.TextInputLayout
```
Comment (gsintrepid) [inline app/src/main/res/layout/fragment_edit_contest_category.xml]: Why do you use TextInputLayout and TextInputEditText?
(Curious because I created a custom view for EditText (with cross button to clear all text) and might incorporate it in your layouts.)

File: app/src/main/res/layout/fragments_invite_contestants.xml [hunk 23]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (jmccormack200) [inline app/src/main/res/layout/fragments_invite_contestants.xml]: Looks like another empty layout.

File: app/src/main/res/values/colors.xml [hunk 24]
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
-    <color name="colorPrimary">#3F51B5</color>
-    <color name="colorPrimaryDark">#303F9F</color>
-    <color name="colorAccent">#FF4081</color>
+    <color name="colorPrimary">#f75f52</color>
+    <color name="colorPrimaryDark">#f54d41</color>
+    <color name="colorAccent">#f75f52</color>
```
Comment (jmccormack200) [inline app/src/main/res/values/colors.xml]: Primary and accent are the same color, is this intentional?

File: app/src/main/res/values/strings.xml [hunk 25]
```diff
@@ -20,4 +20,14 @@
     <string name="pick_image_title">Select Picture</string>
     <string name="entry_image_preview_caption">%1$s looks great!</string>
     <string name="entry_image_content_description">Entry image</string>
+
+    <string name="new_contest">New Contest</string>
```
Comment (jmccormack200) [inline app/src/main/res/values/strings.xml]: I think this may be better as a new_contest_title just for clarity.

File: app/src/main/res/values/strings.xml [hunk 26]
```diff
@@ -20,4 +20,14 @@
...
+    <string name="name_your_contest">Name your contest</string>
+    <string name="scoring_categories">Scoring Categories</string>
+    <string name="category_name">Category Name</string>
+    <string name="description_optional">Description (optional)</string>
+    <string name="error_msg">Error</string>
+    <string name="min_score_weight">0.5x</string>
```
Comment (jmccormack200) [inline app/src/main/res/values/strings.xml]: This is a stretch goal so we dont need these values yet.

File: app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java (hunk 6 above)
Comment (abassawo) [inline app/src/main/java/io/intrepid/contest/screens/contestcreation/NewContestActivity.java]: I was thinking onContestNameEntered() so that it is clear that while it's not a callback, its happening in response to a user input.. whatcha think?
